src/internal/pubsub-client.ts

```diff
@@ -0,0 +1,316 @@
+            truncateString(topicName)
+          );
+        } else if (resp?.discontinuity) {
+          this.logger.trace(
+            'Received discontinuity from subscription stream; topic: %s',
+            truncateString(topicName)
+          );
+        }
+      })
+      .on('error', (err: Error) => {
+        const serviceError = err as unknown as ServiceError;
+        // The service cuts the the stream after ~1 minute. Hence we reconnect.
+        if (
+          serviceError.code === Status.INTERNAL &&
+          serviceError.details === 'Received RST_STREAM with code 0'
+        ) {
+          this.logger.trace(
+            'Server closed stream due to idle activity. Restarting.'
+          );
+          this.sendSubscribe(
+            cacheName,
+            topicName,
+            options,
+            lastTopicSequenceNumber + 1
+          );
+          restartedDueToError = true;
+          return;
+        }
+
+        // Otherwise we propagate the error to the caller.
+        options.errorListener(
+          new TopicSubscribe.Error(cacheServiceErrorMapper(serviceError))
+        );
+      })
+      .on('end', () => {
+        // The stream could have already been restarted due to an error.
+        if (restartedDueToError) {
+          this.logger.trace(
+            'Stream ended after error but was restarted on topic: %s',
+            topicName
+          );
+          return;
+        }
+
+        this.logger.trace('Stream ended on topic: %s; restarting.', topicName);
+        this.sendSubscribe(
+          cacheName,
+          topicName,
+          options,
+          lastTopicSequenceNumber + 1
+        );
+      });
+  }
+
+  private initializeUnaryInterceptors(
+    loggerFactory: MomentoLoggerFactory,
+    middlewares: Middleware[]
+  ): Interceptor[] {
+    return this.initializeInterceptors(
+      loggerFactory,
+      middlewares,
+      this.unaryRequestTimeoutMs
+    );
+  }
+
+  private initializeStreamingInterceptors(
+    loggerFactory: MomentoLoggerFactory,
+    middlewares: Middleware[]
+  ): Interceptor[] {
+    return this.initializeInterceptors(loggerFactory, middlewares);
+  }
+
+  private initializeInterceptors(
+    loggerFactory: MomentoLoggerFactory,
+    middlewares: Middleware[],
+    timeoutMs?: number
+  ): Interceptor[] {
+    const headers = [
+      new Header('Authorization', this.credentialProvider.getAuthToken()),
+      new Header('Agent', `nodejs:${version}`),
+    ];
+    if (timeoutMs) {
+      return [
+        middlewaresInterceptor(loggerFactory, middlewares),
+        new HeaderInterceptorProvider(headers).createHeadersInterceptor(),
+        ClientTimeoutInterceptor(timeoutMs),
+        ...createRetryInterceptorIfEnabled(
+          this.configuration.getLoggerFactory(),
+          this.configuration.getRetryStrategy()
+        ),
+      ];
+    } else {
+      return [
+        new HeaderInterceptorProvider(headers).createHeadersInterceptor(),
+      ];
+    }
```

Comment:
TODO: the unary middlewares do not work with the subscribe streaming response. Figure out what exactly is incompatible, then decide on the correct set of middlewares.

Reply:
can you create a ticket for that and put a comment in the code?

Reply:
https://github.com/momentohq/client-sdk-nodejs/issues/349

Reply:
[a6088c9](https://github.com/momentohq/client-sdk-nodejs/pull/347/commits/a6088c9c6ed245f13d8526c5f44ca6c0b26fbebb)